Refactor Fishsanity

boomshroom opened the pull request:
This makes z_player.c, z_fishing.c, and z_en_fish.c closer to the original decomp by extracting the custom logic into fishsanity.cpp using hooks, VB, etc. It also changes how the fish are organized in location_list.cpp allowing for the rest of the randomizer infrastructure to take care of the details of giving the correct item for the particular check.

In its current form, it removes FastDrops for bottle catches, though they could be added back in using the new VB hook by copying the desired logic from z_player.c along with the bottle catch table. The VB is positioned such that it should allow for anything to be "caught" with a bottle should there be a handler for a particular entity type.

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2004371187.zip)
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2004391369.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2004393348.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2004394279.zip)
<!--- section:artifacts:end -->

boomshroom committed:
Move Fishsanity hooks out of mods.cpp

boomshroom committed:
Assign fish check flags.

boomshroom committed:
Clean up location_list for fish

boomshroom committed:
Prevent fishing from giving double items.

boomshroom committed:
Remove no-longer-used mPendingFish

boomshroom committed:
Override draw function for fishing

This allows the draw functions in the source
overlay to match the decomp.

boomshroom committed:
Override draw function for EnFish

boomshroom committed:
Overwrite grotto fish params based on respawn data

This allows the randomizer to identify them
automatically without any special logic.

The catch (pun not intended) is that grotto fish don't respawn,
and they were previously identified for such by a params value of 1,
so the logic to take care of that needed to be duplicated.
Thankfully it wasn't very much.

boomshroom committed:
Add a VB for catching actors in bottles.

boomshroom committed:
Clean up remaining code after conversion to VB

This breaks fast FastDrops for bottle pickups, though
readding it shouldn't be too hard with the VB hook.

boomshroom committed:
Remove fishsanityParams from Fishing

It was previously used to track exactly which fish would be
released after a catch, but since both candidate fish
would've been caught, they both wouldn't give checks
anyways.

Pepe20129 commented on soh/soh/Enhancements/randomizer/fishsanity.cpp:
```diff
@@ -321,12 +336,171 @@ namespace Rando {
+    }
+
+    void Fishsanity::RegisterHooks() {
+        static uint32_t onActorInitHook = 0;
+        static uint32_t onFlagSetHook = 0;
+        static uint32_t onActorUpdateHook = 0;
+        static uint32_t onSceneInitHook = 0;
+        static uint32_t onVanillaBehaviorHook = 0;
+
+        GameInteractor::Instance->RegisterGameHook<GameInteractor::OnLoadGame>([](int32_t fileNum) {
+            GameInteractor::Instance->UnregisterGameHook<GameInteractor::OnActorInit>(onActorInitHook);
+            GameInteractor::Instance->UnregisterGameHook<GameInteractor::OnFlagSet>(onFlagSetHook);
+            GameInteractor::Instance->UnregisterGameHook<GameInteractor::OnActorUpdate>(onActorUpdateHook);
+            GameInteractor::Instance->UnregisterGameHook<GameInteractor::OnSceneInit>(onSceneInitHook);
+            GameInteractor::Instance->UnregisterGameHook<GameInteractor::OnVanillaBehavior>(onVanillaBehaviorHook);
+
+            onActorInitHook = 0;
+            onFlagSetHook = 0;
+            onActorUpdateHook = 0;
+            onSceneInitHook = 0;
+
+            if (!IS_RANDO || RAND_GET_OPTION(RSK_FISHSANITY) == RO_FISHSANITY_OFF) return;
+
+            FSi->InitializeFromSave();
+
+            onActorInitHook = GameInteractor::Instance->RegisterGameHook<GameInteractor::OnActorInit>(Fishsanity::OnActorInitHandler);
+            onActorInitHook = GameInteractor::Instance->RegisterGameHook<GameInteractor::OnFlagSet>(Fishsanity::OnFlagSetHandler);
+            onActorInitHook = GameInteractor::Instance->RegisterGameHook<GameInteractor::OnActorUpdate>(Fishsanity::OnActorUpdateHandler);
+            onActorInitHook = GameInteractor::Instance->RegisterGameHook<GameInteractor::OnSceneInit>(Fishsanity::OnSceneInitHandler);
+            onActorInitHook = GameInteractor::Instance->RegisterGameHook<GameInteractor::OnVanillaBehavior>(Fishsanity::OnVanillaBehaviorHandler);
+        });
```
I think that it'd be better to integrate this into the main rando `OnLoadGame` hook and not register another `OnLoadGame` hook. I can't come up with a clean way of doing this right now but it'd be great if eventually everything rando related is exclusively in the `OnLoadGame` hook inside `hook_handlers.cpp`.

Pepe20129 commented on soh/soh/Enhancements/randomizer/hook_handlers.cpp:
```diff
@@ -60,7 +61,8 @@ RandomizerCheck GetRandomizerCheckFromFlag(int16_t flagType, int16_t flag) {
                 (flagType == FLAG_EVENT_CHECK_INF && loc.GetCollectionCheck().type == SPOILER_CHK_EVENT_CHK_INF) ||
                 (flagType == FLAG_ITEM_GET_INF && loc.GetCollectionCheck().type == SPOILER_CHK_ITEM_GET_INF) ||
-                (flagType == FLAG_RANDOMIZER_INF && loc.GetCollectionCheck().type == SPOILER_CHK_RANDOMIZER_INF)
+                (flagType == FLAG_RANDOMIZER_INF && loc.GetCollectionCheck().type == SPOILER_CHK_RANDOMIZER_INF) ||
+                (flagType == FLAG_RANDOMIZER_INF && loc.GetCollectionCheck().type == SPOILER_CHK_FISH)
```
This will conflict with #4258 as `SPOILER_CHK_FISH` is removed there as it did the same as `SPOILER_CHK_RANDOMIZER_INF`, in fact (as far as I see), this pr uses `SPOILER_CHK_RANDOMIZER_INF` so there should be no problem removing this.
```suggestion
                (flagType == FLAG_RANDOMIZER_INF && loc.GetCollectionCheck().type == SPOILER_CHK_RANDOMIZER_INF)
```

Pepe20129 commented on soh/soh/Enhancements/randomizer/hook_handlers.cpp:
```diff
@@ -1526,4 +1528,5 @@ void RandomizerRegisterHooks() {
         onSceneInitHook = GameInteractor::Instance->RegisterGameHook<GameInteractor::OnSceneInit>(RandomizerOnSceneInitHandler);
         onActorInitHook = GameInteractor::Instance->RegisterGameHook<GameInteractor::OnActorInit>(RandomizerOnActorInitHandler);
     });
+    Rando::Fishsanity::RegisterHooks();
```
As said in the comment on `fishsanity.cpp`.

Pepe20129 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -13964,38 +13963,10 @@ void Player_Action_8084ECA4(Player* this, PlayState* play) {
-                    Message_StartTextbox(play, D_80854A04[this->av1.actionVar1 - 1].textId, &this->actor);
-                }
+            if (this->av2.actionVar2 == 0) {
+                Message_StartTextbox(play, D_80854A04[this->av1.actionVar1 - 1].textId, &this->actor);
```
I don't see any problem with retaining `FastDrops` here for now.

Pepe20129 commented on soh/src/overlays/actors/ovl_player_actor/z_player.c:
```diff
@@ -14026,43 +13997,14 @@ void Player_Action_8084ECA4(Player* this, PlayState* play) {
-                                this->av1.actionVar1 = -1;
-                                this->av2.actionVar2 = rc;
-                            }
+                            Player_UpdateBottleHeld(play, this, catchInfo->itemId, ABS(catchInfo->itemAction));
```
Same here with `FastDrops`.

Pepe20129 reviewed (commented):
Looks good in general but, there's a couple of things of note.

boomshroom committed:
Update soh/soh/Enhancements/randomizer/hook_handlers.cpp

Co-authored-by: Pepe20129 <72659707+Pepe20129@users.noreply.github.com>

boomshroom committed:
Re-add FastDrops for bottle pickups.

While this does diverge from the vanilla decomp, I'm uncertain of
the order that hooks are run, so I put it back into z_player.c
just to be safe. A future commit can do a more proper implementation
using VB hooks.

boomshroom committed:
Move initialisation of fishsanity hooks into hook_handlers

boomshroom committed:
Merge remote-tracking branch 'origin/develop-rando' into fishsanity_refactor

boomshroom committed:
Merge remote-tracking branch 'origin/develop-rando' into fishsanity_refactor

boomshroom committed:
Change location constructor to take RandomizerInf instead of uint8_t

This shouldn't have an effect as-is, but other changes can add
additional randomizer flags that can end up pushing fishsanity check
flags out of the range of a uint8_t, causing the cast to overflow
and not be stored correctly. With this change, it could still overflow
when writing to the flag field of Location, but said field is unused
and the parameter is really only for setting the flag for the
SpoilerCollectionCheck.

boomshroom committed:
Render uncaught overworld fish as randomized item

Malkierian commented:
Could I get a rebase/pull on dev-rando for this to get builds going again?

boomshroom committed:
Merge remote-tracking branch 'origin/develop-rando' into fishsanity_refactor

boomshroom committed:
Fix windows build by zeroing unused field

boomshroom committed:
Merge remote-tracking branch 'origin/develop-rando' into fishsanity_refactor

boomshroom committed:
Fix scene parameter type

This resolves a build error on Mac and Windows, but Linux
instead buries it in the sea of warnings, meaning I can't
see it until it fails CI.
